### PR TITLE
Add support for filter wheels

### PR DESCRIPTION
**Pull request by AnthonyHorton**
Addresses #768 by adding support for motorised optical filter wheels. The scope of this PR only includes support for the SBIG filter wheels from Diffraction Limited but with the ground work in place it would not be difficult to add support for other filter wheels.

The design follows the pattern established with the `Camera` and `Focuser` classes. There is an `AbstractFilterWheel` base class as well as implementations of the `FilterWheel` class for both the SBIG filter wheels and simulated hardware. Properties, method names and behaviours have been chosen to be consistent with the `Camera` and `Focuser` classes.

~~Flagging as WIP only because there are a few unrelated changes in here (line length fixes & other minor tweaks) that I need to split out into their own PRs. Otherwise~~ I think this is basically ready to go.

### CODE REVIEW

**Comment by AnthonyHorton** [comment]
The Arduino tests are causing trouble again. `test_create_simulator_small_read_buffer` failed, and in the process spat out so many log messages that Travis is refusing to display the whole log.

```
>       report_num = reading['report_num']
E       KeyError: 'report_num'
pocs/tests/test_arduino_io.py:210: KeyError
```

**Comment by wtgee** [comment]
> Unrelated changes now pulled out into #770, #771 & #772. Once those are merged this should be ready for review.

I'll wait for the merge of the others (and maybe remove the WIP if appropriate).

File: pocs/camera/camera.py [hunk 1]
```diff
@@ -498,17 +498,29 @@ def _process_fits(self, file_path, info):
...
 
+        Args:
+            subcomponent (instance of class_name | dict): the subcomponent object, or the keyword
+                arguments required to create it.
+            sub_name (str): name of the subcomponent, e.g. 'focuser'. Will be used as the attribute
+                name, and must also match the name corresponding POCS submodule for this
+                subcomponent, e.g. `pocs.focuser`
+            class_name (str): name of the subcomponent class, e.g. 'Focuser'
+            base_class (class): the base class for the subcomponent, e.g.
+                `pocs.focuser.AbtractFocuser1, used to check whether subcomponent is an instance.
```
Comment (wtgee) [inline pocs/camera/camera.py]: Trailing `1` instead of backtick

File: pocs/camera/sbigudrv.py [hunk 2]
```diff
@@ -317,14 +319,244 @@ def take_exposure(self,
...
+                a Quantity with time units. If a numeric type without units is given seconds will be
+                assumed.
+
+        Raises:
+            RuntimeError: raised if the driver returns an error or if the final status and position
+                are not as expected.
+            pocs.utils.error.Timeout: raised if the move does not end within the period of time
+                specified by the timeout argument.
         """
+        if timeout is not None:
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/camera/sbigudrv.py]: Oh, that looks handy!

File: pocs/filterwheel/sbig.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    def _move_to(self, position, move_event):
+        self._SBIGDriver.cfw_goto(handle=self._handle,
+                                  position=position,
+                                  cfw_event=move_event,
+                                  timeout=self._timeout)
+
+    def _fits_header(self, header):
+        header = super()._fits_header(header)
+        header.set('FW-FW', self.firmware_version, 'Filter wheel firmware version')
```
Comment (wtgee) [inline pocs/filterwheel/sbig.py]: Hm...that's an awkward keyword. ;)

File: pocs/filterwheel/filterwheel.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+class AbstractFilterWheel(PanBase):
+    """
+    Base class for all filter wheels
+
+    Args:
+        name (str, optional): name of the filter wheel
+        model (str, optional): model of the filter wheel
+        camera (pocs.camera.*.Camera, optional): camera that this filter wheel is associated with.
+        filter_names (list of str): names of the filters installed at each filter wheel position
```
Comment (wtgee) [inline pocs/filterwheel/filterwheel.py]: I can imagine a scenario where I want to pass a `dict` of names/positions but might not be necessary.

File: pocs/filterwheel/simulator.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        camera (pocs.camera.*.Camera, optional): camera that this filter wheel is associated with.
+        filter_names (list of str): names of the filters installed at each filter wheel position
+        timeout (u.Quantity, optional): maximum time to wait for a move to complete. Should be
+            a Quantity with time units. If a numeric type without units is given seconds will be
+            assumed. Default is 10 seconds.
+        serial_number (str): serial number of the filter wheel
+        move_time (astropy.units.Quantity, optional): time to move the filter wheel by one position,
+            optional, default 1 second.
+        move_bidirectional (bool, optional): if True will simulate filter wheel which can rotate in
+            either direction, if False (default) will similate a filter wheel that only moves in
```
Comment (wtgee) [inline pocs/filterwheel/simulator.py]: `simulate`

Also, is this only a feature of the simulator? I didn't see anything about this in the actual code.

File: pocs/filterwheel/simulator.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            msg = "Attempt to move filter wheel when already moving"
+            self.logger.error(msg)
+            raise RuntimeError(msg)
+
+        move_distance = position - self.position
+        if move_distance < 0:
+            if not self._move_birectional:
+                move_distance += self._n_positions
+            else:
+                move_distance = -move_distance
```
Comment (wtgee) [inline pocs/filterwheel/simulator.py]: Can you replace the above with a simple `move_distance = position % self._n_positions`?

File: pocs/filterwheel/simulator.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        move_duration = move_distance * self._move_time
+
+        move = threading.Timer(interval=move_duration,
+                               function=self._complete_move,
+                               args=(position, move_event))
+        move.start()
```
Comment (wtgee) [inline pocs/filterwheel/simulator.py]: Just to be safe it seems like this line should go below the next two lines, since they set the condition the `move` timer changes. If for some reason the time got set to zero move time this could feasibly get stuck.

File: pocs/tests/test_filterwheel.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+def test_init(filterwheel):
+    assert isinstance(filterwheel, SimFilterWheel)
+    assert filterwheel.is_connected
+
+
+def test_camera_init():
+    sim_camera = SimCamera(filterwheel={'model': 'simulator',
+                                        'filter_names': ['one', 'deux', 'drei', 'quattro']})
+    assert isinstance(sim_camera.filterwheel, SimFilterWheel)
+    assert sim_camera.filterwheel.is_connected
```
Comment (wtgee) [inline pocs/tests/test_filterwheel.py]: Maybe we want to check these values in the case when there is no filterwheel as well?

File: pocs/camera/sbigudrv.py [hunk 9]
```diff
@@ -317,14 +319,244 @@ def take_exposure(self,
...
+            start_time = time.time()
+            if isinstance(timeout, u.Quantity):
+                timeout = timeout.to(u.second).value
+
+        try:
+            query = self.cfw_query(handle, model)
+            while query['status'] == 'BUSY':
+                if timeout is not None and (time.time() - start_time) > timeout:
+                    msg = "Timeout waiting for filter wheel on {} move to complete".format(handle)
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/camera/sbigudrv.py]: Might be handy. It would be in the preceding log messages, but only if logging at DEBUG level. I'll put it in here too.

File: pocs/tests/test_filterwheel.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    with pytest.raises(AttributeError):
+        filterwheel.filter_names = ["Unsharp mask", "Gaussian blur"]
+
+# Movement
+
+
+def test_move_number(filterwheel):
+    assert filterwheel.position == 1
+    e = filterwheel.move_to(2)
+    assert math.isnan(filterwheel.position)  # position is NaN while between filters
```
Comment (wtgee) [inline pocs/tests/test_filterwheel.py]: Should this move a larger position so there is no chance of some race-condition (not that there probably is anyway)

File: pocs/filterwheel/filterwheel.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        """ Number of positions in the filter wheel """
+        return self._n_positions
+
+    @property
+    def position(self):
+        """ Current integer position of the filter wheel """
+        raise NotImplementedError
+
+    @position.setter
+    def position(self, position):
```
Comment (wtgee) [inline pocs/filterwheel/filterwheel.py]: The setter on the position, while convenient, seems somewhat dangerous to me. I'm not suggesting we do away with it but seems like you could easily accidentally set this.

File: pocs/tests/test_filterwheel.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    e.wait()
+    assert filterwheel.position == 2
+    e = filterwheel.move_to(3, blocking=True)
+    assert e.is_set()
+    assert filterwheel.position == 3
+    filterwheel.position = 4  # Move by assignment to position property blocks until complete
+    assert filterwheel.position == 4
+
+
+def test_move_bad_number(filterwheel):
```
Comment (wtgee) [inline pocs/tests/test_filterwheel.py]: Can you test at `n_positions` as well for both edge cases?

File: pocs/tests/test_filterwheel.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    with pytest.raises(ValueError):
+        filterwheel.move_to(0, blocking=True)  # No zero based numbering here!
+    with pytest.raises(ValueError):
+        filterwheel.position = 99
+
+
+def test_move_name(filterwheel, caplog):
```
Comment (wtgee) [inline pocs/tests/test_filterwheel.py]: Wha?! How did I not know about `caplog` fixture?! That's nice.

File: pocs/camera/sbigudrv.py [hunk 14]
```diff
@@ -317,14 +319,244 @@ def take_exposure(self,
...
+        cfw_info_params = CFWParams(CFWModelSelect[model],
+                                    CFWCommand.GET_INFO,
+                                    CFWGetInfoSelect.FIRMWARE_VERSION)
+        cfw_info_results = CFWResults()
+        with self._command_lock:
+            self._set_handle(handle)
+            self._send_command('CC_CFW', cfw_info_params, cfw_info_results)
+
+        results = self._cfw_parse_results(cfw_info_results)
+        results = {'model': CFWModelSelect(cfw_info_results.cfwModel).name,
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/camera/sbigudrv.py]: It does. The previous line shouldn't be there any more. Discovered in testing that the `GET_INFO` command doesn't return all the same values as the `QUERY` command, despite having fields in the results struct for them, so there's no point in calling `_cfw_parse_results()`

File: pocs/camera/sbigudrv.py [hunk 15]
```diff
@@ -317,14 +319,244 @@ def take_exposure(self,
...
+                asks the driver to autodetect the model.
+
+        Returns:
+            dict: dictionary containing the 'model', 'position', 'status' and 'error' values
+                returned by the driver.
+
+        Raises:
+            RuntimeError: raised if the driver returns an error
+        """
+        cfw_query_params = CFWParams(CFWModelSelect[model],
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/camera/sbigudrv.py]: Fair point. I've pulled out the common stuff into a 6 line helper function.

File: pocs/camera/sbigudrv.py [hunk 16]
```diff
@@ -1289,3 +1533,109 @@ class GetDriverInfoResults0(ctypes.Structure):
     _fields_ = [('version', ctypes.c_ushort),
                 ('name', ctypes.c_char * 64),
                 ('maxRequest', ctypes.c_ushort)]
+
+
+#################################################################################
+# Filter wheel related
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/camera/sbigudrv.py]: The following `Structure`s and `Enum`s are used in this file, and only in this file, so I think it's reasonable for them to be in here.

File: pocs/filterwheel/filterwheel.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+    @property
+    def is_connected(self):
+        """ Is the filterwheel available """
+        return self._connected
+
+    @property
+    def camera(self):
+        """
+        Reference to the Camera object that the FilterWheel is assigned to, if any. A filter wheel
+        should only ever be assigned to one or zero Cameras!
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: The setter for the `FilterWheel` class `camera` property will log a `WARNING` and ignore you if you try to set the `camera` on a filterwheel that already has a `camera`, if that's what you mean.

File: pocs/filterwheel/filterwheel.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+    def position(self, position):
+        self.move_to(position, blocking=True)
+
+    @property
+    def current_filter(self):
+        """ Name of the filter in the current position """
+        try:
+            filter_name = self.filter_names[self.position - 1]  # 1 based numbering
+        except (IndexError, TypeError):
+            # Some filter wheels sometimes cannot return their current position
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: The expectation is that this almost always means that the filter wheel is in between filter positions, hopefully on its way to where it was just asked to move to.

Under those circumstances there's not much that an internal software position tracker could add, and `UNKNOWN` is probably the most appropriate response to the question of 'which filter is currently in position'.

File: pocs/filterwheel/filterwheel.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+            for i, filter_name in enumerate(self.filter_names):
+                if filter_name.startswith(position):
+                    if int_position is None:
+                        int_position = i + 1  # 1 based numbering for filter wheel positions
+                    else:
+                        # Already matched at least once
+                        msg = "More than one filter name matches '{}', using '{}'".format(
+                            position, self.filter_names[int_position - 1])
+                        self.logger.warning(msg)
+                        break
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: Purely so that I can log a warning about multiple matches. Ambigious filter wheel `move_to()` commands probably indicate a mistake somewhere, so shouldn't pass entirely silently.

File: pocs/filterwheel/filterwheel.py [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+        raise NotImplementError
+
+    def move_to(self, position, blocking=False):
+        """
+        Move the filter wheel to the given position.
+
+        The position can be expressed either as an integer, or as (part of) one of the names from
+        the filter_names list. To allow filter names of the form '<filter band>_<serial number>'
+        to be selected by band only position can be a substring from the start of one
+        of the names in the filter_names list, provided that this produces only one match.
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: OK.

File: pocs/filterwheel/filterwheel.py [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+                raise ValueError(msg)
+
+        if int_position < 1 or int_position > self.n_positions:
+            msg = "Position must be between 1 and {}, got {}".format(self.n_positions, int_position)
+            self.logger.error(msg)
+            raise ValueError(msg)
+
+        return int_position
+
+    def _fits_header(self, header):
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: Good point. Will do.

File: pocs/filterwheel/filterwheel.py [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+        header.set('FW-NAME', self.name, 'Filter wheel name')
+        header.set('FW-MOD', self.model, 'Filter wheel model')
```
[the remark this answers is not in the document]
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: Chosen for consistency with `CAM-MOD` and `FOC-MOD` that I already have from the `Camera` and `Focuser` classes.

File: pocs/filterwheel/simulator.py (hunk 5 above)
Comment (AnthonyHorton) [reply pocs/filterwheel/simulator.py]: Yes, it's used to determine how long a simulated move should take, based on whether the filter wheel can move directly to towards the requested position or whether it has to go the long way around because it only rotates in one direction.

Outside of the simulator this is an implementation detail of the filter wheel itself, so the POCS code doesn't need to know about it.

Filter wheels that only rotate in one direction are pretty common, as it improves repeatability. The SBIG filter wheel that I've been testing with is like that, but the only reason I could tell was by watching it.

File: pocs/filterwheel/simulator.py (hunk 6 above)
Comment (AnthonyHorton) [reply pocs/filterwheel/simulator.py]: Sort of...
```
move_distance = position - self.position
if self._move_bidirectional:
    move_distance = abs(move_distance)
else:
    move_distance = move_distance & self._n_positions
```
is equivalent, 1 line shorter, and arguably clearer.

File: pocs/filterwheel/simulator.py (hunk 7 above)
Comment (AnthonyHorton) [reply pocs/filterwheel/simulator.py]: OK. Setting the `Timer` with zero interval does seem to cause trouble, which is one of the reasons I added some code to the higher level `move_to()` method in the base class to prevent the `_move_to()` method from being called for a 'move' to the position that the filter wheel is already in.

File: pocs/filterwheel/filterwheel.py (hunk 4 above)
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: I'm not seeing a need right now. Can add that option later, if it does turn out to be necessary.

File: pocs/tests/test_filterwheel.py (hunk 8 above)
Comment (AnthonyHorton) [reply pocs/tests/test_filterwheel.py]: Well, a camera with no filterwheel will have a `filterwheel` attribute (thanks to the `_create_subcomponent()` call) but the value will be `None`. Adding a test for that.

File: pocs/filterwheel/filterwheel.py (hunk 11 above)
Comment (AnthonyHorton) [reply pocs/filterwheel/filterwheel.py]: It's really just there as a convenience during interactive use (e.g. testing in an IPython shell). The intent is that you almost always use the `.move_to()` method directly so that you can collect the events and move multiple filter wheels in parallel with each other. Perhaps the documentation should say that.

File: pocs/tests/test_filterwheel.py (hunk 13 above)
Comment (AnthonyHorton) [reply pocs/tests/test_filterwheel.py]: It is. I was forced into finding out how to do this because exceptions raised in a `Thread` don't make it out to the calling thread, so the log messages were the only way I could test that the timeout was happening in `test_move_timeout`.

File: pocs/tests/test_filterwheel.py (hunk 12 above)
Comment (AnthonyHorton) [reply pocs/tests/test_filterwheel.py]: OK
